// nuf_messageing/lib/messageing.js

var
    path = require("path"),
    Event = require("nuf_events") 
;

var Messageing = (function(){
    
    function log(message){
        console.log(
            "------ nuf_messageing  /lib/messageing.js  ------\n"+
            message+"\n"+
            "----------------------------------"
        );
    }
    var ClientManager = (function(){
        var KnownClients = (function(){
            function KnownClients(){
                function getNextClientId(){
                    function generateId(){
                        var alpha = "abcdefghjklmnpqrstuvwxyzABCDEFGJKLMNPQRSTUVWXYZ1234567890";
                        var result = "";
                        
                        for(var i=0; i < 47; i++){
                            result += alpha[Math.floor(alpha.length * Math.random()) ];
                        }
                        return result;
                    }
                    
                    var id = generateId();
                    while(clients[id]){
                        id = generateId();
                    }
                    return id;
                }
                function getNewClient(clientManager){
                    var
                        client = new Client(clientManager,getNextClientId())
                    ;
                    Event.addEventEmitter(client);
                    clients[client.id] = client;
                    return client;
                }
                function removeClient(clientId){
                    delete clients[clientId];
                }
                
                function init(){
                    clients = {};
                    _this.getNewClient = getNewClient;
                    _this.removeClient = removeClient;
                    Object.defineProperty(
                        _this,
                        "clients",
                        { get:function(){ return clients;} }
                    );
                    
                }
                var
                    _this = this,
                    clients
                ;
                init();
            }
            return KnownClients;
        })();
        function Client(clientManager, id){
            function dispose(){
                 state = "disposed";
                _this.manager.removeClient(_this.id);
                reqObject.res.end();
            }
            function send(messageObject){
                var
                    trasferString
                ;
                if(state == "connected"){
                    trasferString = MessageObject.getTransferString(messageObject);
                    reqObject.res.write("id: " + Date.now() + "\ndata: " + JSON.stringify(trasferString)+"\n\n");
                }else{
                    throw "Cannot send message!!\nClient:"+id +" is not connected";
                }
            }
            
            function init(){
                Event.addEventEmitter(_this);
                _this.dispose = dispose;
                _this.send = send;
                _this.send.MessageObject = MessageObject;
                
                state = "connecting";
                Object.defineProperty(
                    _this,
                    "id",
                    {
                        get:function(){return id;},
                    }
                );
                Object.defineProperty(
                    _this,
                    "ip",
                    {
                        get:function(){
                            var ip;
                            if(reqObject && reqObject.req){
                                ip = reqObject.req.ip;
                            }
                            return ip;
                        },
                    }
                );
                Object.defineProperty(
                    _this,
                    "reqObject",
                    {
                        get:function(){return reqObject;},
                        set:function(value){
                            reqObject = value;
                            state = "connected";
                        }
                    }
                );
                Object.defineProperty(
                    _this,
                    "state",
                    {
                        get:function(){return state;},
                    }
                );
                Object.defineProperty(
                    _this,
                    "clientManager",
                    {
                        get:function(){return clientManager;},
                    }
                );
            }
            var
                _this = this,
                reqObject,
                state
            ;
            init();
        }
        function ClientManager(){
            function getNewClient(){
                var
                    client = knownClients.getNewClient(_this)
                ;
                client.manager = _this;
                _this.on.dispatchEvent("newClient",{client:client});
                return client;
            }
            function getClient(clientId){
                return knownClients.clients[clientId];
            }
            function getClients(){
                var
                    clients = knownClients.clients,
                    keys = Object.keys(clients), key,
                    i,
                    result = {}
                ;
                for(i =0; i< keys.length; i++){
                    key = keys[i];
                    result[key] = clients[key];
                }
                return result;
            }
            function removeClient(clientId){
                knownClients.removeClient(clientId);
            }
            function init(){
                Event.addEventEmitter(_this);
                knownClients = new KnownClients();
                _this.getNewClient = getNewClient;
                _this.getClient = getClient;
                _this.removeClient = removeClient;
                Object.defineProperty(
                    _this,
                    "clients",
                    {
                        get:function(){ return getClients(); }
                    }
                );
            }
            var
                _this = this,
                knownClients
            ;
            init();
        }
        return ClientManager;
    })();
    var ReqObject = (function(){
        function ReqObject(req, res){
            function sendSuccessfulResponse(responseMessageObject){
                responseMessageObject =
                    responseMessageObject ||
                    new MessageObject(
                        "serverResponse",
                        undefined,
                        "server",
                        {
                            "message":"Success"
                        }
                    );
                res.writeHead(200, {
                    'Content-Type': "application/json",
                    "Cache-Control": "no-cache",
                });
                res.end(MessageObject.getTransferString(responseMessageObject) );
            }
            function sendFailedResponse(responseMessageObject){
                responseMessageObject =
                    responseMessageObject ||
                    new MessageObject(
                        "serverResponse",
                        undefined,
                        "server",
                        {
                            "message":"Failed"
                        },
                        { "error": "Failed"}
                    );
                res.writeHead(400, {
                    'Content-Type': "application/json",
                    "Cache-Control": "no-cache",
                });
                res.end(MessageObject.getTransferString(responseMessageObject) );
            }
            
            function init(){
                _this.req = req;
                _this.res = res;
                _this.sendSuccessfulResponse = sendSuccessfulResponse;
                _this.sendFailedResponse = sendFailedResponse;
            }
            
            var
                _this = this
            ;
            init();
        }
        return ReqObject;
    })();
    var MessageObject = (function(){    
        function MessageObject(type, dest, source, data, flags){
            function stringify(){
                return JSON.stringify({type:type, dest:dest, source:source, data:data, flags:flags});
            }
            function toString(){
                return JSON.stringify(_this, undefined, "\t\t");
            }
            function init(){
                flags = flags || {};
                _this.type = type;
                _this.dest = dest;
                _this.source = source;
                _this.data = data;
                _this.flags = flags;
                _this.stringify = stringify;
                _this.toString = toString;
            }
            var
                _this = this
            ;
            init();
        }
        MessageObject.fromString = function (str){
            var
                json = JSON.parse(str)
            ;
            return new MessageObject(json.type, json.dest, json.source, json.data, json.flags);
        };
        MessageObject.getTransferString = function(messageObject){
            var
                data = messageObject.data,
                headerString,
                dataString = JSON.stringify(data),
                result
            ;
            delete messageObject.data;
            headerString = JSON.stringify(messageObject);
            messageObject.data = data;
            result = headerString+"\n"+ dataString;
            return result;
        };
        return MessageObject;
    })();  
    function Messageing(app, url){
        function onSSEConnect(req, res){
            var
                urlClientId = req.params.clientId,
                client = clientManager.getClient(urlClientId),
                timeout = 1000 * 60* 60 * 24 * 7, // one week timeout
                event
            ;
            if(client){
                client.reqObject = new ReqObject(req, res);
                // set timeout
                req.socket.setTimeout(timeout);
                // client disconnect
                req.on("close", function() {
                    client.on.dispatchEvent("closed", client);
                });
                
                // set headers
                res.writeHead(200, {
                    "Content-Type": "text/event-stream",
                    "Cache-Control": "no-cache",
                    "Connection": "keep-alive"
                });
                event = new Event("connected", {client:client});
                client.on.dispatchEvent(event);
            }else{
                res.writeHead(400);
                res.end();
            }
        }
        function onPostMessage(req, res){
            function ChunkReceiver(){
                function setHeader(){
                    var
                        index
                    ;
                    
                    if(!header && (index = data.indexOf("\n"))> 0){
                        header = MessageObject.fromString(data.substring(0,index) );
                        data = data.substring(index+1);
                        _this.onHeaderLoaded(data);
                    }
                }
                function onData(chunk){
                    function receiveFirstChunk(){
                        chunkCount++;
                        if(chunk.indexOf("blobtype") === 0){
                            dataType = "blobtype";
                        }
                        data = chunk;
                        setHeader();
                    }
                    function stringChunkHandler(){
                        chunkCount++;
                        data += chunk;
                        setHeader();
                    }
                    var chunkHandler = stringChunkHandler;
                    if(chunkCount === 0){
                        receiveFirstChunk();
                    }else{
                        chunkHandler();
                    }
                }
                function getFinalizedData(){
                    header.data = JSON.parse(data);
                    finalizedData = header;
                    return finalizedData;
                }
                function init(){
                    // set interal propertys
                    chunkCount = 0;
                    dataType = "unknown";
                    // set external propertys
                    _this.onData = onData;
                    _this.onHeaderLoaded = function(header){ log("defaut ChunkReceiver.onHeaderLoaded with header:\n"+header);};
                    Object.defineProperty(
                        _this,
                        "rawData",
                        {
                            get:function(){ return data;}
                        }
                    );
                    Object.defineProperty(
                        _this,
                        "finalizedData",
                        {
                            get:function(){
                                var result = finalizedData;
                                if(!result){
                                    result = getFinalizedData();
                                }
                                return result;
                            }
                        }
                    );
                    Object.defineProperty(
                        _this,
                        "chunkCount",
                        {
                            get:function(){ return chunkCount;}
                        }
                    );
                }
                var
                    _this = this,
                    chunkCount,
                    data, finalizedData,
                    dataType,
                    header
                ;
                init();   
            }
            function getMessageObject(){
                var chunkReceiver  = new ChunkReceiver();
                req.setEncoding('utf8');
                req.on('data', chunkReceiver.onData);
                req.on(
                    'end',
                    function(){
                        var
                            messageObject = chunkReceiver.finalizedData,
                            reqObject = new ReqObject(req, res),
                            urlClientId = req.params.clientId
                        ;
                        if(messageObject.constructor.name == "MessageObject"){
                            _this.on.dispatchEvent(
                                "message",
                                {
                                    messageObject:messageObject,
                                    urlClientId:urlClientId,
                                    reqObject:reqObject
                                }
                            );
                        }else{
                            _this.on.dispatchEvent(
                                "textMessage",
                                {
                                    textMessage:messageObject,
                                    urlClientId:urlClientId,
                                    reqObject : new ReqObject(req, res)
                                }
                            );
                        }
                    }
                );
            }
            getMessageObject();
        }
        function appInit(app){
            var
                preUrlStr = "/"+((url[url.length-1] === "/")?url.substring(0,url.length-1):url),
                postUrl = preUrlStr+"/post/:clientId?",
                sseUrl = preUrlStr+"/sseClientConnect/:clientId?",
                messageingClientUrl = preUrlStr+"/messageingClient.js",
                messageingClientPath = path.resolve(__dirname+"/../src/messageingClient.js")
            ;
            app.get (sseUrl,onSSEConnect);
            app.post(postUrl,onPostMessage);
            app.use (messageingClientUrl, require("express").static(messageingClientPath) );
            Event.appInit(app);
        }
        function defaultOnMessage(event){
            log(
                "messaging default onMessage received an unhandled messageObject.\n"+
                "event:"+event
            );
        }
        function defaultOnTextMessage(event){
            log(
                "messaging default onTextMessage received an unhandled text message.\n"+
                "event:"+event
            );
        }
        
        function init(){
            Event.addEventEmitter(_this);
            url = url || "/";
            clientManager = new ClientManager();
            appInit(app);
            _this.onMessage = defaultOnMessage;
            _this.onTextMessage = defaultOnTextMessage;
            // external properties
            Object.defineProperty(
                _this,
                "clientManager",
                 {
                    get:function(){ return clientManager;}
                 }
            );
            Object.defineProperty(
                _this,
                "app",
                 {
                    get:function(){ return app;}
                 }
            );
        }
        var
            _this = this,
            clientManager            
        ;
        init();
    }
    Messageing.MessageObject = MessageObject;
    Messageing.blobToBase64 = function(blob, callback){
        reader.readAsDataURL(blob); 
        reader.onloadend = function() {
            callback(reader.result);
        };
    };
    Messageing.base64TobBlob = function(b64String, callback){
        fetch(url)
        .then(res => res.blob())
        .then(blob => callback(blob));
    };
    Messageing.arraybufferToBase64 = function(arrayBuffer, callback){
        var blob = new Blob([arrayBuffer],{type:'application/octet-binary'});
        blobToBase64(blob, callback);
    };
    Messageing.arraybufferToBase64 = function(arrayBuffer, callback){
        var blob = new Blob([arrayBuffer],{type:'application/octet-binary'});
        blobToBase64(blob, callback);
    };
    //WorkerUtilities.blobToBase64;
    //Messageing.arrayBufferToBase64 = WorkerUtilities.blobToBase64;
    //Messageing.base64ToBlob = WorkerUtilities.base64ToBlob;
    //Messageing.base64ToArrayBuffer = WorkerUtilities.base64ToArrayBuffer;
    return  Messageing;
})();

module.exports = Messageing;
